Introduce BackendService for managing external storage backends

RobinMcCorkell opened the pull request:
Backends are registered to the BackendService through new data structures:

BackendConfig specifies the basic information for a backend: class, human-readable name, visibility to personal/admin, initial priority. It takes an array of BackendParameter's to specifiy the parameters

BackendParameter stores a parameter configuration for an external storage: name of parameter, human-readable name, type of parameter (text, password, hidden, checkbox), flags (optional or not).

When a BackendConfig has checkDependencies() called on it, it will return a (possibly empty) array of BackendDependency's that specify a dependency of the backend. If an empty array is returned it means there are no unhandled dependencies, so the backend can be used.

Storages in the StoragesController now get their parameters validated server-side (fixes a TODO).

TODO:
- [x] Unit tests
- [x] Fix the Local storage type appearing in 'allow user mounts' (but cannot be selected)
- [x] Ask devs if the API is good
- [x] Integrate hooks (at least login)
- [x] Convert all existing storages to authentication methods
- [x] Migrate SMB_OC to SMB + OC auth method
- [x] Migrate OpenStack to OpenStack + Rackspace separate authentication methods
- [x] Check custom JS backends still work
- [x] Check custom JS auth methods work
- [x] Check multi-auth method backends allow all to be configured

cc @PVince81 @icewind1991 @DeepDiver1975

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c207486e-63df-4e68-b4c3-58822ed9b350) was created.

DeepDiver1975 commented on apps/files_external/lib/backendconfig.php:
```diff
@@ -0,0 +1,260 @@
+/**
+ * External storage backend configuration
+ */
+class BackendConfig implements \JsonSerializable {
```
@PVince81 shall we move this into core? In the long run files_external will die and continue to live as isolated apps.
The UI and necessary APIs have to be in core.

PVince81 replied:
Yes, makes sense.

Xenopathic replied:
@DeepDiver1975 That was one of the main reasons for implementing this - moving BackendService and StorageService over to core is much easier than trying to port OC_Mount_Config over. We just need to make sure we get the API right - that's what the [WIP] tag is for! :smile:

nickvergessen commented on apps/files_external/service/backendservice.php:
```diff
@@ -0,0 +1,340 @@
+			->setHasDependencies(true)
+		);
+
+		if (!\OC_Util::runningOnWindows()) {
```
:kill it: its not supported anymore :see_no_evil:

Xenopathic replied:
I was never quite sure of this - so any `runningOnWindows()` check can just be removed, indiscriminately?

DeepDiver1975 replied:
> :kill it: its not supported anymore :see_no_evil:

hold your horses - in case somebody will come back to take care of Windows support - that one will be happy to find these code pieces in here.

To remove any windows specific code - if at all - is a tasks of it's own.

PVince81 commented:
CC @jmaciasportela as this will be useful for future work on external storage

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/018f5272-f8d4-4035-b91a-2703d68f3d56) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4db24211-aa66-4e2e-814b-34fc25bd57ae) was created.

Xenopathic commented:
This is becoming quite large... if anyone wants to comment on the API, please do so! I'm trying to avoid too many inter-class dependencies, while still providing a clean interface that doesn't involve passing class names everywhere.

Tests are quite comprehensive at this point...

LukasReschke commented on apps/files_external/lib/auth/imechanism.php:
```diff
@@ -0,0 +1,42 @@
+	 *
+	 * @param array $params
+	 */
+	public function __construct($params);
```
`array $params`

LukasReschke commented on apps/files_external/lib/auth/nullmechanism.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * @param array $params
+	 */
+	public function __construct($params) {
```
`array $params`

LukasReschke commented on apps/files_external/lib/auth/password/basic.php:
```diff
@@ -0,0 +1,49 @@
+	/**
+	 * @param array $params
+	 */
+	public function __construct($params) {
```
`array $params`

LukasReschke commented on apps/files_external/lib/backendconfig.php:
```diff
@@ -0,0 +1,288 @@
+	 * @param string $text Human-readable name
+	 * @param BackendParameter[] $parameters
+	 */
+	public function __construct($class, $text, $parameters) {
```
`array $parameters`

PVince81 commented on apps/files_external/lib/auth/nullmechanism.php:
```diff
@@ -0,0 +1,46 @@
+ * Null authentication mechanism
+ * Implements AuthMechConfig::SCHEME_NULL
+ */
+class NullMechanism implements IMechanism {
```
Not sure if it's worth having a whole interface just for `toParams()`. Could almost be a simple JSON serializable or something.

Or should this rather be a base class instead of interface ?

PVince81 replied:
I'd usually expect such interface to have a method "authenticate()" or something.
It seems so far it's just a container for parameters ?

Maybe thinking about the more complex case could help figuring the interface out ? How would the other auths work ?

PVince81 replied:
(I'm actually talking about `IMechanism`, not the null one :smile:)

PVince81 commented on apps/files_external/templates/settings.php:
```diff
@@ -14,103 +18,146 @@
+				</td>
+				<td class="configuration">
+					<?php $options = $storage->getBackendOptions(); ?>
+					<?php foreach ($storage->getBackend()->getParameters() as $parameter): ?>
```
Argh. I hope one day we can render the full list in JS alone, not in PHP.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/09959c99-7c47-4b65-85e6-923c30b5a3dd) was created.

PVince81 commented on apps/files_external/lib/backendparameter.php:
```diff
@@ -0,0 +1,179 @@
+	const VALUE_TEXT = 0;
+	const VALUE_BOOLEAN = 1;
+	const VALUE_PASSWORD = 2;
+	const VALUE_HIDDEN = 3;
```
We might want another type `VALUE_CUSTOM = 4`.
A backend should be then be able to specify a custom type name, and then provide JS code to render and handle it.
This could be done separately, but keep it in mind in case it affects the way how params are rendered. I guess the PHP could only render a placeholder "div" and then trigger an event somehow to trigger the custom JS code to render it.

PVince81 replied:
One example could be a dropdown or multiselect. Not sure what other kind of parameters we need, @jmaciasportela do you know ?

Xenopathic replied:
I don't see how a custom value constant here would be useful - do you mean creating a new placeholder prefix (eww) that is exposed to JS when that value type is set? Or that backends can *specify* the placeholder prefix they want to use (ewwwwww)?

PVince81 replied:
Imagine if there was a time travelling storage where you can specify a date. That one would want to have a date picker, which is currently not possible. So it would specify the custom flag so no field would be rendered, but a value would still be read/written to the config. But maybe that could be achieved with a simple text field, which would be automatically replaced with a date picker by the custom JS code.

Xenopathic replied:
Ahh, I see, so no rendering of an input at all! Brilliant! /me scurries away to change it

PVince81 replied:
Just noticed we have the "HIDDEN" one, maybe that one will already do.
So never mind.

Xenopathic replied:
Actually on second thoughts, it's not going to work very well. A custom field still needs positioning in the configuration area, so there needs to be some kind of placeholder element that the custom JS can replace, preserving the placement of the inputs. I think the easiest way to accomplish this is via the `VALUE_HIDDEN` type, since that won't be rendered yet still 'allocates' a space in the configuration area.

Thoughts?

PVince81 replied:
Whether space is allocated isn't that bad. With jquery you can just insert elements anywhere.
Most important is that the value is properly read/written in the field.
If using a hidden field, the JS code for that field should make sure to update the value in the hidden field so it can be saved properly. (similar to how we do mount options)

Xenopathic replied:
Well, it's not necessary to keep the field as a hidden field - the custom JS can replace the element entirely. The advantage of having an element to start with is that the ordering of the fields is guaranteed, and no magic is required to figure out how to insert between the `URL` and `Root` fields for example :smile:

PVince81 commented:
We might need a fallback `OC_Mount_Config::registerBackend()` for older / third party ext storage implementations. I'd consider this method public API even though the naming doesn't match.
The implementation would just be an adapter from the old format to the new one.
Then add `@deprecated` so we can remove it later.

PVince81 commented on apps/files_external/service/backendservice.php:
```diff
@@ -0,0 +1,394 @@
+
+use \OCA\Files_External\Lib\BackendConfig;
+use \OCA\Files_External\Lib\AuthMechConfig;
+use \OCA\Files_External\Lib\BackendParameter as Param;
```
Remove `as Param` ? I didn't find any occurrence

PVince81 commented on apps/files_external/service/backendservice.php:
```diff
@@ -0,0 +1,394 @@
+
+		$this->registerBackend(
+			(new BackendConfig('\OC\Files\Storage\Local', $l->t('Local'), [
+				(new Param('datadir', $l->t('Location'))),
```
ah here... never mind

PVince81 replied:
I still suggest to rename it here to BackendParam, because people might copy paste this for their own backends :smile:

Xenopathic replied:
I was trying to be conservative for my keyboard :laughing:

PVince81 commented on apps/files_external/service/backendservice.php:
```diff
@@ -0,0 +1,394 @@
+				(new Param('use_path_style', $l->t('Enable Path Style')))
+					->setType(Param::VALUE_BOOLEAN),
+			]))
+			->setHasDependencies(true)
```
How about `setDependenciesCheckFunction()` and pass a callback that checks for deps ?

Xenopathic commented:
Pluggable authentication mechanisms are here!

![capture](https://cloud.githubusercontent.com/assets/2016878/8908094/f4e1437c-3471-11e5-85e4-669c642c0deb.PNG)

The 'Static password' mechanism is just a quick hack I threw together, which always returns the same password (in this case, the correct one for the user). Note that the password *never* gets stored in `mount.json`!

Selecting a different authentication mechanism from the dropdown re-creates the relevant configuration fields.

Next up: working with hooks. Some authentication mechanisms might want to listen for hooks (login, namely) and perform some action, such as updating `mount.json`, as a result. This will be tough without using static methods...

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/0ac31baf-928d-4f03-9444-a6b0e21e8236) was created.

Xenopathic commented:
@PVince81 Awww, I hoped to avoid compatibility shims in `OC_Mount_Config` - eventually I want to kill that class completely.

PVince81 commented:
We all do :wink: 
The problem is that without shim we can't merge your PR until all external storage backends have been adapted, even stuff like irods... so better keep compatible for now.

Xenopathic commented:
/me sees 'iRODS', runs away.

Xenopathic commented:
Actually to be fair, I stole my original implementation idea of SMB_OC from iRODS, so it had *some* use!

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/834f2d0c-1fd6-4c11-ab4a-2998af6395a1) was created.

Xenopathic commented:
@owncloud-bot retest this please

Xenopathic commented:
15:00:13 PhantomJS 1.9.8 (Linux 0.0.0) ERROR
    15:00:13   SyntaxError: Parse error
    15:00:13   at /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_external/js/settings.js:9

Line 9 is the beginning `(function() {`, and running JSHint + some other checkers locally doesn't flag any errors...

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e5298da5-08bd-40c3-b9a6-294526b90c67) was created.

Xenopathic commented:
@owncloud-bot retest this please

Xenopathic commented:
Still getting that JS parse error... any ideas @PVince81 ?

PVince81 commented:
@Xenopathic forget about ES6 and remove the default parameter on line 760

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f7058438-62f8-4689-aec7-e2eef669d2b1) was created.

Xenopathic commented:
@PVince81 So I changed the way auth mechanisms work again... see that last commit: https://github.com/owncloud/core/commit/764e588a683cc2c2dfb2d789a703321dea786cc8

> An AuthMechConfig now takes a fully instantiated implementation, instead
of the class + scheme. This allows an implementation to have
dependencies injected into it, e.g. an \OCP\IDB for database operations.
The downside is that authentication mechanisms now 'look' different to
backends, which is an inconsistency.
>
> When a StorageConfig is being prepared for mounting, the authentication
mechanism implementation (IMechanism) has manipulateStorage() called,
which inserts the relevant authentication method options into the
storage ready for mounting.
>
> An AuthMechConfig is still required to define the frontend
configuration, such as parameters, dependencies and custom JS files.

I did think of a way to make everything consistent, and more powerful as a result. Scrap the concept of BackendConfig and AuthMechConfig, and have a class for each backend that needs to be registered (separate from the actual storage class OFC). So, for example, `\OCA\Files_External\Backends\SMB`, which implements `\OCA\Files_External\Backends\IBackend`. It has the duty of defining the parameters, localised name, auth schemes, dependencies, custom JS etc, basically what BackendConfig does at the moment. But with a fully instantiated separate class, instead of a 'config' class. Do the same for AuthMechConfig, just with `IMechanism` instead of `IBackend`.

Advantages:
 * One class per backend/auth mechanism
 * No fake 'config' classes
 * Dependency injection on everything - a backend could do fancy work before the actual storage is created
 * A better concept of 'backend' and 'auth mechanism'
 * Use of interfaces everywhere - more powerful OOP
 * Both backends and auth mechanisms can listen to hooks/events easily
 * Backends themselves can be unit tested

Disadvantages:
 * More classes (=== (slightly) slower code)
 * Apps using the old `registerBackend()` method may need reimplementing

If we are doing this, we might as well do it properly (and *before* we merge with core and have to support the APIs in OCP)

Thoughts? cc @icewind1991 @LukasReschke @MorrisJobke @DeepDiver1975

PVince81 commented on apps/files_external/lib/authmechconfig.php:
```diff
@@ -0,0 +1,157 @@
+use \OCA\Files_External\Lib\Auth\IMechanism;
+
+/**
+ * External storage authentication mechanism configuration
```
looks like it's more than just config if it contains custom JS + implementation ?

PVince81 replied:
The naming might be confusing. From the impl this seems to only be the configuration definition, not the configuration values (where StorageConfig contains the real values). I can't think of a better name though.

PVince81 commented on apps/files_external/lib/authmechconfig.php:
```diff
@@ -0,0 +1,157 @@
+	private $text;
+
+	/** @var BackendParameter[] parameters for mechanism */
+	private $parameters = [];
```
so basically these are just the parameter definitions, not the parameter values

PVince81 replied:
and these parameters are actually the ones that exist also in StorageConfig ? (inferring from the validation method)

PVince81 commented on apps/files_external/lib/auth/imechanism.php:
```diff
@@ -0,0 +1,49 @@
+	 *
+	 * @param StorageConfig $storage
+	 */
+	public function manipulateStorage(StorageConfig &$storage);
```
=> updateStorageConfig()

PVince81 replied:
Ok I guess we don't need an "authenticate" method as this can be done here.
For example in the case of SSO if we need to retrieve a token from a remote server, this could be done inside this method, correct ?

Xenopathic replied:
Well, the idea is we don't constrain auth mechanisms by the method names they use for hooks and the like. Since an auth mechanism now comes with a proper implementation, that is created during registration, in the constructor it can call `connectHook()` and attach hooks to whatever methods it likes.

PVince81 commented:
@Xenopathic already looking better than last time, making great progress :smile:

Xenopathic commented:
To see what the 'fully instantiated' version would look like, aka the one without any BackendConfig or AuthMechConfig at all, have a look at the occ console command registration. A new command extends the Symfony Console Command class, which is created when registered and then to actually be executed the `execute()` method is called. I propose a similar thing here, for the backends and auth mechanisms.

Xenopathic commented:
- [ ] BUG: Personal storages cannot override global ones

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/de2fe0f2-a0da-4853-8e64-126095e6e7aa) was created.

Xenopathic commented:
Unit tests will fail until #18095 is resolved

Xenopathic commented:
- [ ] Unavailable auth mechanism/backend information should have a special exception
- [ ] Unavailable storages via auth mechanism/backend should register a 'fake' MountPoint that always throws StorageNotAvailableException
- [ ] 'Status unknown' in config UI, in case information not currently available (e.g. session credentials)
- [ ] Create auth mechanism that saves credentials to the config/DB (might leave this one for a future PR, this one is big enough)

Xenopathic commented:
- [x] Merge SMB with SMB_OC
- [ ] Merge SFTP with SFTP_Key

Xenopathic commented:
- [ ] SMB_OC should not appear in backend selection lists (but should still be able to be mounted)
- [ ] SFTP_Key should not appear in backend selection lists (but should still be able to be mounted)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/25d68f19-259d-4909-926d-a02a573094f1) was created.

Xenopathic commented:
Nearly there. The following still needs doing:

- [ ] SMB_OC should not appear in backend selection lists (but should still be able to be mounted)
- [ ] SFTP_Key should not appear in backend selection lists (but should still be able to be mounted)
- [ ] Unavailable auth mechanism/backend information should have a special exception
- [ ] Unavailable storages via auth mechanism/backend should register a 'fake' MountPoint that always throws StorageNotAvailableException, so storage unavailability is correctly propagated
- [ ] 'Status unknown' in config UI, in case information not currently available (e.g. session credentials)
- [ ] Migrate JS client side to relevant auth mechanisms, e.g. sftp_key JS file now belongs as part of the publickey auth scheme, and should work with any backend that supports that scheme.

PVince81 commented:
@Xenopathic it's fine to have basic credentials as default auth without JS UI. The JS UI can be adjusted later.
I'd say for the first version have credential auth by default and not changeable. Goal is to save some time and be able to merge this PR as soon as possible with a minimal viable backward-compatible solution.

Xenopathic commented:
Since this PR is so huge, here's a review matrix for the important parts. Please review bits when you have time, and edit the matrix with your name!

- Lib\DefinitionParameter | Lib\MissingDependency
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Lib\*Trait
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Lib\Backend\Backend
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Lib\Backend\* : specific backends
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Lib\Auth\AuthMechanism
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Lib\Auth\* : specific mechanisms
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Lib\Config : legacy compatibility
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Service\BackendService
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Service\UserGlobalStoragesService
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:

Specific changes

- SMB backend changes
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- SFTP backend changes
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:
- Unit test coverage
- [ ] 1 Reviewed by:
- [ ] 2 Reviewed by:

Testing

- [ ] Tested by:
- [ ] Tested by:
- [ ] Tested by:
- ...

Xenopathic commented:
@PVince81 The thing is, the JS is currently completely broken, since I've changed the way backends are referenced. The goal is that any existing configured storages 'just work' on upgrade to this version. The PHP side should be working, it's just the JS that will break.

PVince81 commented:
You can try making the JS work, but don't add anything related to auth type selection yet.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/26c50694-3e79-45a6-9cbf-6035bc8aeeac) was created.

MorrisJobke commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/09236b81-4ab2-468c-b0c0-b46e4b1c31c3) was created.

Xenopathic commented:
I do believe this is just about finished. I expect there's plenty of corner cases, especially wi

Xenopathic commented:
Oops, pressed the wrong button. Still getting used to my new laptop!

DeepDiver1975 commented:
Wow - this is a rather big PR! Any chance to get this slices up into multiple PRs? This is no reviewable.

THX

Xenopathic commented:
@DeepDiver1975 Not without causing far more hassle than is necessary. The later commits build on the earlier ones, so it really isn't feasible to split everything up. Please see the review matrix I posted (https://github.com/owncloud/core/pull/15914#issuecomment-128745327), as that provides a good template to work on. Review by parts is how it's going to have to be.

Alternatively, a thorough test is worth a thousand reviews (not quite, but it's probably close!).

PVince81 commented:
@Xenopathic please set to "To Review" and remove the "WIP" if you think this is ready.
I see you still have unchecked boxes here, please update: https://github.com/owncloud/core/pull/15914#issue-71578403

Xenopathic commented:
@PVince81 Just working on JS now, then I'll set for review.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/28f4b915-b42b-4251-9cc6-168d2436b4f9) was created.

Xenopathic commented:
Time to review!

jospoortvliet commented:
Out of curiosity, as I'll be writing about this at some point, how does this, if at all, relate to https://github.com/owncloud/core/pull/12086

PVince81 commented:
@jospoortvliet this has nothing to do with shared storage refactor https://github.com/owncloud/core/pull/12086 which itself is a completely different topic.

Xenopathic commented:
@jospoortvliet I'm not familiar with the sharing code, but as far as I can tell thisis completely separate.

jospoortvliet commented:
ok, sorry for the noise

DeepDiver1975 commented on apps/files_external/appinfo/routes.php:
```diff
@@ -47,10 +47,10 @@
-	->actionInclude('files_external/ajax/dropbox.php');
-$this->create('files_external_google', 'ajax/google.php')
-	->actionInclude('files_external/ajax/google.php');
+$this->create('files_external_oauth1', 'ajax/oauth1.php')
```
what is the reasoning behind this rename from dropbox to oauth?
dropbox has it's own mechanism and we are done. I don't expect this auth mechanism to work with any other external storage (and even if - :sleeping: : )

Xenopathic replied:
Dropbox uses the OAuth1 mechanism, and with the appropriate changes to oauth1.php it could support any backend that needs OAuth1 support, not just Dropbox.

DeepDiver1975 commented on apps/files_external/js/dropbox.js:
```diff
@@ -1,111 +0,0 @@
-$(document).ready(function() {
```
I would have loved an approach with simply keeps these files untouched

Xenopathic replied:
See oauth1.js :smile:

DeepDiver1975 commented:
What about introducing the new interfaces/services on php side first and then make up our mind on how to move the storage implementation over?

Any ui related stuff should for sure be don on its own.

Finally we should really only touch ONE storage at first - e.g. dav - otherwise QA will go nuts

DeepDiver1975 commented:
> Finally we should really only touch ONE storage at first - e.g. dav - otherwise QA will go nuts

the untouched storages should continue to live as is

Xenopathic commented:
@DeepDiver1975 Could you elaborate on what you mean by 'touch' the storages? The actual implementations (aka lib/dropbox.php, lib/sftp.php) are unchanged with the exception of SFTP (merged public key auth), the only new bit is the registration of the backends (lib/backend/*.php)

PVince81 commented on apps/files_external/lib/failedstorage.php:
```diff
@@ -0,0 +1,200 @@
+/**
+ * Storage placeholder to represent a missing precondition, storage unavailable
+ */
+class FailedStorage extends Common {
```
Not so sure about this. Was this necessary ?

Xenopathic replied:
Yes, so that any failures in getting information before the MountPoint is created (e.g. no session credentials available) are passed through to the filesystem layer, and clients are correctly notified of the storage unavailability. Otherwise if there is an auth error or something clients will start to delete files as they think the storage has disappeared.

PVince81 commented on apps/files_external/lib/identifiertrait.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Trait for objects requiring an identifier (and/or identifier aliases)
+ */
+trait IdentifierTrait {
```
Seems this is used only in a single location ?
How about just moving these directly into that class ? (no premature optimization/factorizing)

Xenopathic replied:
Two, Backend and AuthMechanism

PVince81 commented on apps/files_external/lib/auth/authmechanism.php:
```diff
@@ -0,0 +1,98 @@
+	const SCHEME_PUBLICKEY = 'publickey';
+	const SCHEME_OPENSTACK = 'openstack';
+
+	use VisibilityTrait;
```
auth mechanisms also have visibility ?

Xenopathic replied:
Yes. Currently not used for admin configuration, but in the future you may want an auth mechanism to only be visible to the admin, say, or to be deprecated so no new storages can be created with it. That is all handled by the VisibilityTrait.

DeepDiver1975 replied:
> Yes. Currently not used for admin configuration, but in the future you may want an auth mechanism to only be visible to the admin, say, or to be deprecated so no new storages can be created with it. That is all handled by the VisibilityTrait.

we should really only add what is immediately required.

DeepDiver1975 commented:
> @DeepDiver1975 Could you elaborate on what you mean by 'touch' the storages? The actual implementations (aka lib/dropbox.php, lib/sftp.php) are unchanged with the exception of SFTP (merged public key auth), the only new bit is the registration of the backends (lib/backend/*.php)

the js scripts for google and dropbox have been touched -> requires full retest which has a bigger implication

Xenopathic commented:
@DeepDiver1975 AH, I see.

OK, the plan is as follows:

1. Cut out the changes to nearly all backends/auth mechanisms, leaving only the DAV class and Username and password auth mechanism + the API
2. Introduce backend conversions with additional auth mechanisms in future PR(s)

The API should now be stable though - get reviewing!

PVince81 commented on apps/files_external/templates/settings.php:
```diff
@@ -14,103 +70,123 @@
+							});
+							/*uasort($authMechanisms, function($a, $b) {
+								return strcasecmp($a->getText(), $b->getText());
+							});*/
```
old code?

PVince81 commented:
The API looks fine!

Commit:
Introduce BackendService for managing external storage backends

Backends are registered to the BackendService through new data
structures:

Backends are concrete classes, deriving from
\OCA\Files_External\Lib\Backend\Backend. During construction, the
various configuration parameters of the Backend can be set, in a design
similar to Symfony Console.

DefinitionParameter stores a parameter configuration for an external
storage: name of parameter, human-readable name, type of parameter
(text, password, hidden, checkbox), flags (optional or not).

Storages in the StoragesController now get their parameters validated
server-side (fixes a TODO).

Commit:
Load custom JS for all backends/auth mechanisms

Loading custom JS on a per-backend basis added needless complexity and
made dealing with async required. Now all backends/auth mechanisms load
custom JS in PHP

Commit:
Compatibility shims for OC_Mount_Config

The following functions have been removed:
 - addMountPoint()
 - removeMountPoint()
 - movePersonalMountPoint()

registerBackend() has been rewritten as a shim around BackendService,
allowing legacy code to interact with the new API seamlessly

addMountPoint() was already disconnected from all production code, so
this commit completes the job and removes the function itself, along
with disconnecting and removing related functions. Unit tests have
likewise been removed.

getAbsoluteMountPoints(), getSystemMountPoints() and
getPersonalMountPoints() have been rewritten to use the StoragesServices

Commit:
Unit tests for new backend API

Commit:
Authentication mechanisms for external storage backends

A backend can now specify generic authentication schemes that it
supports, instead of specifying the parameters for its authentication
method directly. This allows multiple authentication mechanisms to be
implemented for a single scheme, providing altered functionality.

This commit introduces the backend framework for this feature, and so at
this point the UI will be broken as the frontend does not specify the
required information.

Terminology:
 - authentication scheme
    Parameter interface for the authentication method. A backend
    supporting the 'password' scheme accepts two parameters, 'user' and
    'password'.
 - authentication mechanism
    Specific mechanism implementing a scheme. Basic mechanisms may
    forward configuration options directly to the backend, more advanced
    ones may lookup parameters or retrieve them from the session

New dropdown selector for external storage configurations to select the
authentication mechanism to be used.

Authentication mechanisms can have visibilities, just like backends.
The API was extended too to make it easier to add/remove visibilities.
In addition, the concept of 'allowed visibility' has been introduced, so
a backend/auth mechanism can force a maximum visibility level (e.g.
Local storage type) that cannot be overridden by configuration in the
web UI.

An authentication mechanism is a fully instantiated implementation. This
allows an implementation to have dependencies injected into it, e.g. an
\OCP\IDB for database operations.

When a StorageConfig is being prepared for mounting, the authentication
mechanism implementation has manipulateStorage() called,
which inserts the relevant authentication method options into the
storage ready for mounting.

Commit:
Introduce UserGlobalStoragesService

UserGlobalStoragesService reads the global storage configuration,
cherry-picking storages applicable to a user. Writing storages through
this service is forbidden, on punishment of throwing an exception.
Storage IDs may also be config hashes when retrieved from this service,
as it is unable to update the storages with real IDs.

As UserGlobalStoragesService and UserStoragesService share a bit of code
relating to users, that has been split into UserTrait. UserTrait also
allows for the user set to be overridden, rather than using the user
from IUserSession.

Config\ConfigAdapter has been reworked to use UserStoragesService and
UserGlobalStoragesService instead of
OC_Mount_Config::getAbsoluteMountPoints(), further reducing dependance
on that horrible static class.

Commit:
Propagate auth mechanism/backend failures to filesystem layer

Failure to prepare the storage during backend or auth mechanism
manipulation will throw an InsufficientDataForMeaningfulAnswerException,
which is propagated to StorageNotAvailableException in the filesystem
layer via the FailedStorage helper class.

When a storage is unavailable not due to failure, but due to
insufficient data being available, a special 'indeterminate' status is
returned to the configuration UI.

Commit:
Split backend identifiers from the class name

Prior to this, the storage class name was stored in mount.json under the
"class" parameter, and the auth mechanism class name under the
"authMechanism" parameter. This decouples the class name from the
identifier used to retrieve the backend or auth mechanism.

Now, backends/auth mechanisms have a unique identifier, which is saved in
the "backend" or "authMechanism" parameter in mount.json respectively.
An identifier is considered unique for the object it references, but the
underlying class may change (e.g. files_external gets pulled into core
and namespaces are modified).

Commit:
Make Application a singleton

The same Application must be used in the settings templates and in
routes, so that any registered backends are correctly seen

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/00c101e5-9397-4f61-8fbb-81c67c5c6a54) was created.

Xenopathic commented:
OK, there are now no backends in this PR. Also, I've squashed a load of the commits, so there is a lot less iterative history.

That's the last time I try to rebase 29 commits, rearrange them and squash them down...

PVince81 commented:
- [ ] BUG: mount SFTP on master, switch to this branch: storage appears red in GUI, but still mounted properly

PVince81 commented:
- [ ] BUG: cannot select any auth than "None" for new storages. Please make "credentials" available as default

PVince81 commented:
I also see this in the log:
```
{"reqId":"bphI\/+ZCa86uwW9iI04N","remoteAddr":"127.0.0.1","app":"index","message":"Exception: {\"Exception\":\"UnexpectedValueException\",\"Message\":\"Invalid backend \\\\OC\\\\Files\\\\Storage\\\\SFTP\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/service\\\/storagesservice.php(225): OCA\\\\Files_external\\\\Service\\\\StoragesService->populateStorageConfigWithLegacyOptions(Object(OCA\\\\Files_external\\\\Lib\\\\StorageConfig), 'user', 'all', Array)\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/service\\\/storagesservice.php(306): OCA\\\\Files_external\\\\Service\\\\StoragesService->readConfig()\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/controller\\\/storagescontroller.php(219): OCA\\\\Files_external\\\\Service\\\\StoragesService->getStorage(1)\\n#3 [internal function]: OCA\\\\Files_External\\\\Controller\\\\StoragesController->show(1)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/appframework\\\/http\\\/dispatcher.php(159): call_user_func_array(Array, Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/appframework\\\/http\\\/dispatcher.php(89): OC\\\\AppFramework\\\\Http\\\\Dispatcher->executeController(Object(OCA\\\\Files_External\\\\Controller\\\\GlobalStoragesController), 'show')\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/appframework\\\/app.php(108): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OCA\\\\Files_External\\\\Controller\\\\GlobalStoragesController), 'show')\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/appframework\\\/routing\\\/routeactionhandler.php(45): OC\\\\AppFramework\\\\App::main('GlobalStoragesC...', 'show', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#8 [internal function]: OC\\\\AppFramework\\\\routing\\\\RouteActionHandler->__invoke(Array)\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(274): call_user_func(Object(OC\\\\AppFramework\\\\routing\\\\RouteActionHandler), Array)\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/base.php(879): OC\\\\Route\\\\Router->match('\\\/apps\\\/files_ext...')\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/index.php(40): OC::handleRequest()\\n#12 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/service\\\/storagesservice.php\",\"Line\":86}","level":4,"time":"2015-08-13T10:21:41+00:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files_external\/globalstorages\/1"}
```

PVince81 commented:
Hmm I think I understand the problem. Since you didn't adapt SFTP it defaults to "None". I think we need to make the legacy storages default to "Basic credentials". I'd even suggest to hide the auth dropdown if there is only a single auth. ("No auth" selection doesn't make much sense)

PVince81 commented:
- [ ] BUG: but not reproducible, maybe the stack trace will help find out which situation it could happen:
```
{"reqId":"VE5dURz+a4ioQEo4HQd6","remoteAddr":"127.0.0.1","app":"index","message":"Exception: {\"Exception\":\"DomainException\",\"Message\":\"UserGlobalStoragesService writing disallowed\",\"Code\":0,\"Trace\":\"
#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/service\\\/globalstoragesservice.php(95): OCA\\\\Files_External\\\\Service\\\\UserGlobalStoragesService->writeLegacyConfig(Array)\\n
#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/service\\\/storagesservice.php(246): OCA\\\\Files_external\\\\Service\\\\GlobalStoragesService->writeConfig(Array)\\n
#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/service\\\/storagesservice.php(321): OCA\\\\Files_external\\\\Service\\\\StoragesService->readConfig()\\n
#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/lib\\\/config\\\/configadapter.php(114): OCA\\\\Files_external\\\\Service\\\\StoragesService->getAllStorages()\\n
#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/config\\\/mountprovidercollection.php(61): OCA\\\\Files_External\\\\Config\\\\ConfigAdapter->getMountsForUser(Object(OC\\\\User\\\\User), Object(OC\\\\Files\\\\Storage\\\\StorageFactory))\\n
#5 [internal function]: OC\\\\Files\\\\Config\\\\MountProviderCollection->OC\\\\Files\\\\Config\\\\{closure}(Object(OCA\\\\Files_External\\\\Config\\\\ConfigAdapter))\\n
#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/config\\\/mountprovidercollection.php(62): array_map(Object(Closure), Array)\\n
#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/filesystem.php(425): OC\\\\Files\\\\Config\\\\MountProviderCollection->getMountsForUser(Object(OC\\\\User\\\\User))\\n
#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/filesystem.php(350): OC\\\\Files\\\\Filesystem::initMountPoints('root')\\n
#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/util.php(198): OC\\\\Files\\\\Filesystem::init('root', '\\\/root\\\/files')\\n
#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/base.php(878): OC_Util::setupFS()\\n
#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/index.php(40): OC::handleRequest()\\n
#12 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files_external\\\/service\\\/userglobalstoragesservice.php\",\"Line\":91}","level":4,"time":"2015-08-13T10:09:23+00:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files\/?dir=%2Fsftp"}
```

I actually had an existing SFTP storage from a previous test, coming from OC 8->8.1->master->this branch. But I couldn't reproduce this any more.

Xenopathic commented:
@PVince81 Yes, all backend migration has been moved to the other PR, so this one just uses the legacy backend registration. As such, there is no authentication mechanisms available.

Perhaps instead of calling it 'None', the authentication mechanism should be called 'Built-in' ?

PVince81 commented:
"Built-in" or "Default" sounds good. But best would be to just hide the dropdown whenever there is only a single choice, to avoid confusion.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/110497f2-a28d-48e0-9fc8-c99897ed20b4) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/badfc477-f793-4157-a8cb-b43d52d36d7f) was created.

Xenopathic commented:
I haven't implemented the 'no dropdown if there's only one auth mechanism' thing you suggested, since the cleanest way to do that is to render the storages list in JS rather than from the template. But that's a whole other PR...

Xenopathic commented:
@PVince81 I'm getting a bunch of JS errors like this:

```
PhantomJS 1.9.8 (Linux 0.0.0) OCA.External.Settings tests storage configuration selecting backend populates the row and creates a new empty one FAILED
        TypeError: 'null' is not an object (evaluating 'e.length')
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:4
            at /srv/local/owncloud/core/apps/files_external/js/settings.js:9
            at /srv/local/owncloud/core/core/vendor/underscore/underscore.js:701
            at /srv/local/owncloud/core/core/vendor/underscore/underscore.js:716
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:5
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:5
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:5
            at /srv/local/owncloud/core/core/vendor/jquery-migrate/jquery-migrate.min.js:2
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:5
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:4
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:4
            at /srv/local/owncloud/core/core/vendor/jquery/jquery.min.js:5
            at selectBackend (/srv/local/owncloud/core/apps/files_external/tests/js/settingsSpec.js:83)
            at /srv/local/owncloud/core/apps/files_external/tests/js/settingsSpec.js:96
```

PhantomJS isn't giving very useful backtraces though, as I can't see what's wrong or where the failing line is.

PVince81 commented:
@Xenopathic try opening http://localhost:9876/debug.html and the JS debugger. Set a breakpoint in /srv/local/owncloud/core/apps/files_external/js/settings.js:9
I guess `this._allAuthMechanism` is null for some reason.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f48ad8b9-3487-48cc-ba2f-44d3e7f929c8) was created.

Xenopathic commented:
@PVince81 You were right, the authentication mechanisms weren't being loaded into the test DOM. I wish Karma would keep each of the real JS file separate, so the backtrace actually has some useful line number information.

Xenopathic commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/9c3adbab-8f70-4d01-99b9-4eee3c6455e7) was created.

Xenopathic commented:
The bug with storage ID updating in UserGlobalStoragesService has been fixed, so any storages retrieved via that service will not result in an ID update, instead leaving them with the config hash as an ID. This is OK, since UserGlobalStoragesService is a read-only layer anyway, so the ID is pretty useless.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/644477b2-c07c-493d-87e9-548705f55619) was created.

Xenopathic commented:
@owncloud-bot retest this please

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/3905de03-d9fb-4b07-bd12-e16d3e41437f) was created.

Xenopathic commented:
OK, everything should now be working. Please review! I've tested the basic backends, the custom JS backends (as far as I can tell it works as before), and the more complex SMB_OC.

It might be a good idea to review this commit-by-commit, since it separates everything into neat chunks.

cc @icewind1991 @MorrisJobke @PVince81 @DeepDiver1975 @LukasReschke

Xenopathic commented:
@owncloud-bot retest this please

icewind1991 commented on apps/files_external/appinfo/application.php:
```diff
@@ -45,5 +49,33 @@ public function __construct(array $urlParams=array()) {
+	 */
+	protected function loadBackends() {
+		$container = $this->getContainer();
+		$service = $container->query('OCA\\Files_External\\Service\\BackendService');
```
does this method actually do anything?

Xenopathic replied:
Not in this PR - see the migration of backends PR. I left it here since I wanted to make the rebasing a bit easier.

icewind1991 commented on apps/files_external/lib/config.php:
```diff
@@ -152,146 +111,71 @@ public static function initMountPointsHook($data) {
-				if (!isset($options['priority'])) {
-					$options['priority'] = $backends[$options['class']]['priority'];
-				}
+		$userGlobalStoragesService->setUser($user);
```
Passing the user as argument to `getAllStorages` would be a cleaner approach

Xenopathic replied:
Agreed. I'll have a look at how easy it will be to have the user passed in per-method instead of in the constructor of User(Global)StoragesService

Xenopathic replied:
Hmm, the current approach passes IUserSession into the constructor of the services, changing that to passing an IUser will be awkward and involve lots of parameter passing. I'm going to leave it as it is for now, if there is a use case for changing it then a future PR will be needed

PVince81 commented:
Code looks good, let's retest this.

icewind1991 commented on apps/files_external/lib/personalmount.php:
```diff
@@ -20,15 +20,33 @@
+	 * @param UserStoragesService $storagesService
+	 * @param int $storageId
+	 */
+	public function attachStoragesService(UserStoragesService $storagesService, $storageId) {
```
why not inject this in the constructor?

Xenopathic replied:
At the time of writing it I had a couple of different approaches. After settling to this one, I forgot to change it to constructor-based initialization. Fixing.

PVince81 commented:
Tests:
- [x] TEST: legacy storage from OC 7 can be upgraded to this branch
- [x] TEST: applicable field
- [x] TEST: rename mount point in settings
- [x] TEST: rename personal mount point in file list
- [x] TEST: mount options from OC 8.0 are kept

- [ ] TODO: for a separate PR: when only one auth backend available, either hide the column or render it has static text

- [ ] BUG: checkout stable8.1, setup ext storage, upgrade to this branch with the web UI: blank page (works fine with CLI upgrade)

PVince81 commented:
- [ ] BUG: can't delete files from ext storages:
```
{"reqId":"sSSbeqOvy5KMg34KrMeq","remoteAddr":"127.0.0.1","app":"PHP","message":"Call to undefined method OC_Mount_Config::getSystemMountPoints() at \/srv\/www\/htdocs\/owncloud\/lib\/private\/encryption\/util.php#293","level":3,"time":"2015-08-17T13:37:21+00:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/delete.php"}
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/52ef4c17-accc-4696-b6c8-67b7bf5b9669) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c6f1df4e-0bb5-4536-ad56-469e043aa160) was created.

Xenopathic commented:
@PVince81 I introduced some compat shims for `getSystemMountPoints()` and `getPersonalMountPoints()`, please test if that fixes your encryption issue.

PVince81 commented:
I didn't have any encryption issues?!

Xenopathic commented:
@PVince81 The issue with deleting a storage, resulting in a problem in the encryption utilities. :smile:

PVince81 commented:
Now getting this on delete:
```
{"reqId":"STYveslOvscjPkqpRfdy","remoteAddr":"127.0.0.1","app":"PHP","message":"Access to undeclared static property: OC_Mount_Config::$appContainer at \/srv\/www\/htdocs\/owncloud\/apps\/files_external\/lib\/config.php#162","level":3,"time":"2015-08-17T14:51:06+00:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/delete.php"}
```

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/59a643fa-ee29-42ff-8e32-1784d9c1e14d) was created.

Xenopathic commented:
@PVince81 Forgot about the Application singleton changes. Try now

Xenopathic commented:
@owncloud-bot retest this please

PVince81 commented:
The blocking ticket https://github.com/owncloud/core/issues/18109 was merged.

This is good to go :+1:

MorrisJobke commented on apps/files_external/appinfo/routes.php:
```diff
@@ -28,8 +28,7 @@
  **/
-$application = new Application();
-$application->registerRoutes(
+\OC_Mount_Config::$app->registerRoutes(
```
This causes an error on my side.
```
{"reqId":"\/ic4rLtLHtT1DuBLflnU","remoteAddr":"127.0.0.1","app":"PHP","message":"Class 'OC_Mount_Config' not found at \/home\/mjob\/Projekte\/owncloud\/master\/apps\/files_external\/appinfo\/routes.php#31","level":3,"time":"2015-08-19T08:05:22+00:00","method":"GET","url":"\/master\/index.php"}
```

I guess this is caused by the order of loaded files. The class file is hardcoded in the app.php.

MorrisJobke replied:
Somehow this worked just very well for a long time, until I reset my whole instance and enabled the files_external from this branch without having the old one enabled.

PVince81 replied:
You might need to merge with master to get the other fix.

MorrisJobke replied:
:white_check_mark:

Xenopathic replied:
Fixed with rebase

MorrisJobke commented:
* [ ] BUG: adding a local mount via personal page I got following error:

```
{"reqId":"oQb7DRZkUfcunqWk0Sy+","remoteAddr":"127.0.0.1","app":"PHP","message":"Argument 1 passed to OCA\\Files_External\\Lib\\PersonalMount::__construct() must be an instance of OCA\\Files_external\\Service\\UserStoragesService, instance of OC\\Files\\Storage\\Local given, called in \/home\/mjob\/Projekte\/owncloud\/master\/apps\/files_external\/lib\/config\/configadapter.php on line 148 and defined at \/home\/mjob\/Projekte\/owncloud\/master\/apps\/files_external\/lib\/personalmount.php#49","level":0,"time":"2015-08-19T08:13:01+00:00","method":"GET","url":"\/master\/index.php"}
{"reqId":"oQb7DRZkUfcunqWk0Sy+","remoteAddr":"127.0.0.1","app":"PHP","message":"Object of class OC\\Files\\Storage\\StorageFactory could not be converted to string at \/home\/mjob\/Projekte\/owncloud\/master\/lib\/private\/files\/filesystem.php#795","level":0,"time":"2015-08-19T08:13:01+00:00","method":"GET","url":"\/master\/index.php"}
{"reqId":"oQb7DRZkUfcunqWk0Sy+","remoteAddr":"127.0.0.1","app":"PHP","message":"Call to undefined method OCA\\Files_External\\Lib\\PersonalMount::attachStoragesService() at \/home\/mjob\/Projekte\/owncloud\/master\/apps\/files_external\/lib\/config\/configadapter.php#149","level":3,"time":"2015-08-19T08:13:01+00:00","method":"GET","url":"\/master\/index.php"}
```


```
$ cat data/mjob/mount.json
{
    "user": {
        "mjob": {
            "\/mjob\/files\/abc": {
                "id": 1,
                "backend": "\\OC\\Files\\Storage\\Local",
                "authMechanism": "builtin::builtin",
                "options": {
                    "datadir": "\/home\/mjob\/tm"
                },
                "storage_id": "3"
            }
        }
    }
}
```

PVince81 commented:
@MorrisJobke weird, you shouldn't be able to select "Local" in the personal page.

MorrisJobke commented:
> @MorrisJobke weird, you shouldn't be able to select "Local" in the personal page.

I also wondered about it, because it was not available some time ago.

Xenopathic commented:
@PVince81 @MorrisJobke Unfortunately due to the way backend registration now works, all legacy storages are (by default) available to everyone. I'll patch in a quick fix for that, but it'll be ugly and will touch the legacy registration code a bit.

Commit:
Restrict Local backend to admin-only

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f5e32c6d-f0af-4117-b9db-9a41e76cc241) was created.

Xenopathic commented:
@MorrisJobke That bug you found was caused by me changing PersonalMount in response to @icewind1991 's comment, but then forgetting to change the point that it actually gets used. Fixed now.

Xenopathic commented:
@owncloud-bot retest this please

Xenopathic commented:
@PVince81 Could you quickly review that last commit I pushed? Everything else is the same.

PVince81 commented:
:+1:

PVince81 commented:
@MorrisJobke retry?

MorrisJobke commented:
Tested and works so far:+1:

Xenopathic commented:
plzdontbreak plzdontbreak plzdontbreak :see_no_evil:
